### DISCUSSION TITLE
How to fully ignore whitespace

### DISCUSSION THREAD

**Discussion by NicViviers** [pest-parser/pest#643] [General]
My grammar does not rely on whitespace, and from the book I am unable to understand how handling whitespace works.
From what I see I think `rule ~ rule` automatically matches whitespace, however I'm getting errors when working from that assumption. The problem:
```
decl = _{ "struct" ~ ident ~ "{" ~ "}" }
```
Does not match:
```
struct String {}
```

It produces the following error:
```
Err(Error { variant: ParsingError { positives: [ident], negatives: [] }, location: Pos(6), line_col: Pos((1, 7)), path: None, line: "struct String {}", continued_line: None })
```

I have no need for whitespace, it can simply be ignored. Anybody able to help me out with this?

**Comment by CAD97** [comment]
You want to define what your grammar's [implicit whitespace](https://pest.rs/book/grammars/syntax.html#implicit-whitespace) is. e.g.

```pest
WHITESPACE = _{ " " | NEWLINE }
```

**Reply by NicViviers** [reply]
I've even tried this but with `"\n"` instead of `NEWLINE`, my mistake then.
I appreciate the fast response. Thank you
